libcrux-ml-kem/src/vector/portable/serialize.rs

```diff
@@ -1,324 +1,416 @@
+// called serialize_N_int or deserialize_N_int that (de)serializes
+// the minimal number of inputs K such that N*K is a multiple of 8.
+// These functions are then called multiple times in the main function,
+// called serialize_N or deserialize_N.
+// This refactoring reduces redundancy, and also makes the code easier for
+// F* to handle. As a general rule, any function that modifies an array
+// more than 8 times with complex expressions starts to strain F*, so
+// we separate out the code that does the computation (in _int functions)
+// and code that updates arrays (in the outer functions).
+
 use super::vector_type::*;
 use crate::vector::traits::FIELD_ELEMENTS_IN_VECTOR;
 
 #[inline(always)]
 pub(crate) fn serialize_1(v: PortableVector) -> [u8; 2] {
     let mut result = [0u8; 2];
-
     for i in 0..8 {
         result[0] |= (v.elements[i] as u8) << i;
     }
-
     for i in 8..16 {
         result[1] |= (v.elements[i] as u8) << (i - 8);
     }
-
     result
 }
 
 #[inline(always)]
 pub(crate) fn deserialize_1(v: &[u8]) -> PortableVector {
     let mut result = zero();
-
     for i in 0..8 {
         result.elements[i] = ((v[0] >> i) & 0x1) as i16;
     }
     for i in 8..FIELD_ELEMENTS_IN_VECTOR {
         result.elements[i] = ((v[1] >> (i - 8)) & 0x1) as i16;
     }
-
     result
 }
 
+#[inline(always)]
+pub(crate) fn serialize_4_int(v: &[i16]) -> (u8, u8, u8, u8) {
+    let result0 = ((v[1] as u8) << 4) | (v[0] as u8);
+    let result1 = ((v[3] as u8) << 4) | (v[2] as u8);
+    let result2 = ((v[5] as u8) << 4) | (v[4] as u8);
+    let result3 = ((v[7] as u8) << 4) | (v[6] as u8);
+    (result0, result1, result2, result3)
+}
+
 #[inline(always)]
 pub(crate) fn serialize_4(v: PortableVector) -> [u8; 8] {
+    let result0_3 = serialize_4_int(&v.elements[0..8]);
+    let result4_7 = serialize_4_int(&v.elements[8..16]);
     let mut result = [0u8; 8];
-
-    result[0] = ((v.elements[1] as u8) << 4) | (v.elements[0] as u8);
-    result[1] = ((v.elements[3] as u8) << 4) | (v.elements[2] as u8);
-    result[2] = ((v.elements[5] as u8) << 4) | (v.elements[4] as u8);
-    result[3] = ((v.elements[7] as u8) << 4) | (v.elements[6] as u8);
-
-    result[4] = ((v.elements[8 + 1] as u8) << 4) | (v.elements[8 + 0] as u8);
-    result[5] = ((v.elements[8 + 3] as u8) << 4) | (v.elements[8 + 2] as u8);
-    result[6] = ((v.elements[8 + 5] as u8) << 4) | (v.elements[8 + 4] as u8);
-    result[7] = ((v.elements[8 + 7] as u8) << 4) | (v.elements[8 + 6] as u8);
-
+    result[0] = result0_3.0;
+    result[1] = result0_3.1;
+    result[2] = result0_3.2;
+    result[3] = result0_3.3;
+    result[4] = result4_7.0;
+    result[5] = result4_7.1;
+    result[6] = result4_7.2;
+    result[7] = result4_7.3;
     result
 }
 
+#[inline(always)]
+pub(crate) fn deserialize_4_int(bytes: &[u8]) -> (i16, i16, i16, i16, i16, i16, i16, i16) {
+    let input = u32::from_le_bytes(bytes.try_into().unwrap());
```

Comment:
This generates noise in the C code because it introduces an explicit panic. It's also more code. So I don't think this is a great style. And accessing the `bytes` directly is still not more than 8 operations and it's what's being done in the other functions too.

Reply:
We actually had two versions of this function, one in `portable/serialize.rs` and another in `neon/serialize.rs`. This version was in `neon` and in my benchmarks seemed to be faster, so I picked that one. But benchmarks are noisy, and I have now reverted to using byte arrays here. In the future, if (de)serialization becomes a bottleneck we can look at it again.